syntax = "proto3";

package ds_homework2;

message Sensor {
    int32 uid = 1;
}

message SensorInfoResponse {
    enum SensorState {
        UP = 0;
        DOWN =  1;
    }

    string name = 1;
    SensorState state = 2;
    float latitude = 3;
    float longitude = 4;
    float altitude = 5;
}

message TemperatureData {
    float temperature = 1;
}

service TemperatureSensor {
    rpc SensorInfo(Sensor) returns (SensorInfoResponse);

    rpc SensorData(Sensor) returns (stream TemperatureData);
}

message Image {
    int32 width = 1;
    int32 height = 2;
    bytes image_data = 3;
    int32 frame_id = 4;
}

message BoundingBox {
    int32 x = 1;
    int32 y = 2;
    int32 width = 3;
    int32 height = 4;
}

message Detection {
    BoundingBox bounding_box = 1;
    int32 frame_id = 2;

    enum Category {
        Person = 0;
        Car = 1;
        Truck = 2;
    }

    Category category = 3;
}

message Detections {
    repeated Detection list = 1;
}

message SummaryResponse {
    int32 num_people = 1;
    int32 num_cars = 2;
    int32 num_trucks = 3;
}

service ObjectDetection {
    rpc SummaryDetection(stream Image) returns (SummaryResponse);
    rpc Detect(stream Image) returns (stream Detections);
}